system-addon/content-src/lib/init-store.js

```diff
@@ -44,12 +44,25 @@ const messageMiddleware = store => next => action => {
 };
 
 const rehydrationMiddleware = store => next => action => {
+  if (!store._earlyMessageQueue) {
+    store._earlyMessageQueue = [];
+  }
+
   if (store._didRehydrate) {
     return next(action);
   }
 
   const isMergeStoreAction = action.type === MERGE_STORE_ACTION;
   const isRehydrationRequest = action.type === at.NEW_TAB_STATE_REQUEST;
+  const isSessionPerfAction = action.type === at.SAVE_SESSION_PERF_DATA;
+
+  // Treasure up the session perf actions to the early message queue if the hydration
+  // is not done yet. All those actions will be dispatched on the MERGE_STORE_ACTION,
+  // which is guaranteed to come from chrome.
+  if (!store._didRehydrate && isSessionPerfAction) {
+    store._earlyMessageQueue.push(action);
+    return null;
```

Comment:
I don't quite understand what's expected from the return value of middleware action handling… @k88hudson ?